rustworkx-core/src/coloring.rs

```diff
@@ -220,8 +221,218 @@ where
     edge_colors
 }
 
-#[cfg(test)]
+struct MisraGries<G: GraphBase> {
+    // The input graph
+    graph: G,
+    // Partially assigned colors (indexed by internal edge index)
+    colors: Vec<Option<usize>>,
+}
+
+impl<G> MisraGries<G>
+where
+    G: EdgeIndexable + IntoEdges,
+{
+    pub fn new(graph: G) -> Self {
+        let colors = vec![None; graph.edge_bound()];
+        MisraGries { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: G::NodeId) -> HashSet<usize> {
+        let used_colors: HashSet<usize> = self
+            .graph
+            .edges(u)
+            .filter_map(|edge| self.colors[EdgeIndexable::to_index(&self.graph, edge.id())])
+            .collect();
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: G::NodeId) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let free_color: usize = (0..)
+            .position(|color| !used_colors.contains(&color))
+            .unwrap();
+        free_color
+    }
+
+    // Returns true iff color c is free at node u
+    fn is_free_color(&self, u: G::NodeId, c: usize) -> bool {
+        let used_colors = self.get_used_colors(u);
+        !used_colors.contains(&c)
+    }
+
+    // Returns the maximal fan on edge (u, v) at u
+    fn get_maximal_fan(
+        &self,
+        eid: G::EdgeId,
+        u: G::NodeId,
+        v: G::NodeId,
+    ) -> Vec<(G::EdgeId, G::NodeId)> {
+        let mut fan: Vec<(G::EdgeId, G::NodeId)> = Vec::new();
+        fan.push((eid, v));
```

Comment:
```suggestion
        let mut fan: Vec<(G::EdgeId, G::NodeId)> = vec![(eid, v)];
```

Reply:
Done in [7900916](https://github.com/Qiskit/rustworkx/pull/902/commits/7900916402fb101ad39724e2d0509619ce10907a). Also the code has been updated to use `G::EdgeRef` instead of `G::EdgeId`, so it's enough to only keep the edge reference.